File: leetcode/0200-number-of-islands.cc
// https://leetcode.com/problems/number-of-islands/
// Medium

#include<vector>

using namespace std;

class Solution {
    int width;
    int height;
    void markVisited(vector<vector<char>>& grid, int i, int j) {
        if (i < 0 || i >= height || j < 0 || j >= width || grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        markVisited(grid, i+1, j);
        markVisited(grid, i-1,j);
        markVisited(grid, i, j+1);
        markVisited(grid, i, j-1);

    }
public:
    int numIslands(vector<vector<char>>& grid) {
        width = grid[0].size();
        height = grid.size();
        int islands = 0;
        for (int i = 0; i < height; ++i) {
            for (int j = 0; j < width; ++j) {
                if (grid[i][j] == '1') {
                    islands++;
                    markVisited(grid, i, j);
                }
            }
        }
        return islands;
    }
};

File: leetcode/python/0200-number-of-islands.py
# https://leetcode.com/problems/number-of-islands/
# Medium

# Simple DFS
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        moves = [[0,1],[1,0],[-1,0],[0,-1]]
        
        def removeOnes(i: int, j: int):
            nonlocal grid

            if i >= 0 and j >= 0 and j < n and i < m and grid[i][j] == '1':
                grid[i][j] = '0'
                for move in moves:
                    removeOnes(i+move[0], j+move[1])
                    
        
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    count += 1
                    removeOnes(i, j)
        return count